### PR TITLE
Integrate inferStream with endpoint

### PR DESCRIPTION
**Pull request by dkalinowski**
JIRA: CVS-122330

### CODE REVIEW

File: src/test/mediapipeflow_test.cpp [hunk 1]
```diff
@@ -975,6 +1006,402 @@ TEST_P(MediapipeFlowAddTest, Infer) {
...
+}
+
+TEST_P(MediapipeFlowAddTest, InferStreamDisconnectionBeforeFirstRequest) {
+    const ovms::Module* grpcModule = server.getModule(ovms::GRPC_SERVER_MODULE_NAME);
+    KFSInferenceServiceImpl& impl = dynamic_cast<const ovms::GRPCServerModule*>(grpcModule)->getKFSGrpcImpl();
+
+    MockedServerReaderWriter<::inference::ModelStreamInferResponse, ::inference::ModelInferRequest> stream;
+    EXPECT_CALL(stream, Read(::testing::_))
+        .WillOnce([](::inference::ModelInferRequest* req) {
+            return false;  // immediate disconnection
```
Comment (mzegla) [inline src/test/mediapipeflow_test.cpp:1398]: This word "immediate" makes me wonder whether this test will be prone to sporadic racing conditions etc.

File: src/kfs_frontend/kfs_grpc_inference_service.cpp [hunk 2]
```diff
@@ -326,6 +325,27 @@ Status KFSInferenceServiceImpl::ModelInferImpl(::grpc::ServerContext* context, c
...
+        Status status = StatusCode::MEDIAPIPE_UNINITIALIZED_STREAM_CLOSURE;
+        SPDLOG_DEBUG(status.string());
+        return status;
+    }
+    std::shared_ptr<MediapipeGraphExecutor> executor;
+    auto status = this->modelManager.createPipeline(executor, firstRequest.model_name(), &firstRequest, nullptr /* response not present in streaming api */);
+    if (!status.ok()) {
+        return status;
+    }
+    return executor->inferStream(firstRequest, *stream);
```
[the remark this answers is not in the document]
Comment (dkalinowski) [reply src/kfs_frontend/kfs_grpc_inference_service.cpp:342]: Stream gets invalidated only when Read returns false, meaning it has disconnected

File: src/test/mediapipeflow_test.cpp [hunk 3]
```diff
@@ -975,6 +1006,402 @@ TEST_P(MediapipeFlowAddTest, Infer) {
...
+        startReloading.get_future().get();
+        MediapipeGraphConfig mgc{
+            modelName,
+            "",                                               // default base path
+            "/ovms/src/test/mediapipe/graphscalar_tf.pbtxt",  // graphPath - valid but includes missing models, will fail for new streams
+            "",                                               // default subconfig path
+            ""                                                // dummy md5
+        };
+        auto status = definition->reload(modelManager, mgc);
+        ASSERT_EQ(status, StatusCode::OK) << status.string();
```
[the remark this answers is not in the document]
Comment (dkalinowski) [reply src/test/mediapipeflow_test.cpp:1323]: Bewlo we check that new stream can create new graph. And that the graph has an error - meaning it is new definition of the graph. Therefore we are sure that the graph definition has reloaded and is available.

File: src/test/mediapipeflow_test.cpp (hunk 1 above)
Comment (dkalinowski) [reply src/test/mediapipeflow_test.cpp:1398]: This test is fully deterministic, meaning it expects that "Read" will be called only once Write will be never called.

**Review by mzegla — APPROVED** [review]

File: src/test/mediapipeflow_test.cpp [hunk 4]
```diff
@@ -975,6 +1006,402 @@ TEST_P(MediapipeFlowAddTest, Infer) {
...
+                ASSERT_EQ(outputs[0].shape().size(), 2);
+                ASSERT_EQ(outputs[0].shape()[0], 1);
+                ASSERT_EQ(outputs[0].shape()[1], 10);
+            }();
+            // expect first response
+            checkAddResponse("out", requestData1[0], requestData1[0], request[0], msg.infer_response(), 1, 1, modelName);
+            return true;
+        })
+        .WillOnce([requestData1, &request, modelName](const ::inference::ModelStreamInferResponse& msg, ::grpc::WriteOptions options) {
+            [&msg]() {
```
Comment (atobiszei) [inline src/test/mediapipeflow_test.cpp:1066]: Calling checkAddResponse isn't sufficient?

File: src/test/mediapipeflow_test.cpp [hunk 5]
```diff
@@ -975,6 +1006,402 @@ TEST_P(MediapipeFlowAddTest, Infer) {
...
+            return true;  // correct sending 3rd request
+        })
+        .WillOnce([](::inference::ModelInferRequest* req) {
+            return false;  // disconnection
+        });
+    EXPECT_CALL(stream, Write(::testing::_, ::testing::_))
+        .WillOnce([requestData1, &request, modelName](const ::inference::ModelStreamInferResponse& msg, ::grpc::WriteOptions options) {
+            [&msg]() {
+                const auto& outputs = msg.infer_response().outputs();
+                ASSERT_EQ(outputs.size(), 1);
```
Comment (atobiszei) [inline src/test/mediapipeflow_test.cpp:1055]: Calling checkAddResponse isn't sufficient?

File: src/test/mediapipeflow_test.cpp [hunk 6]
```diff
@@ -975,6 +1006,402 @@ TEST_P(MediapipeFlowAddTest, Infer) {
...
+                ASSERT_EQ(outputs[0].shape()[0], 1);
+                ASSERT_EQ(outputs[0].shape()[1], 10);
+            }();
+            // expect second response
+            checkAddResponse("out", requestData1[1], requestData1[1], request[1], msg.infer_response(), 1, 1, modelName);
+            return true;
+        })
+        .WillOnce([requestData1, &request, modelName](const ::inference::ModelStreamInferResponse& msg, ::grpc::WriteOptions options) {
+            [&msg]() {
+                const auto& outputs = msg.infer_response().outputs();
```
Comment (atobiszei) [inline src/test/mediapipeflow_test.cpp:1080]: Calling checkAddResponse isn't sufficient?

File: src/test/mediapipeflow_test.cpp (hunk 4 above)
Comment (dkalinowski) [reply src/test/mediapipeflow_test.cpp]: Good catch, fixed

File: src/test/mediapipeflow_test.cpp (hunk 5 above)
Comment (dkalinowski) [reply src/test/mediapipeflow_test.cpp]: Fixed

File: src/test/mediapipeflow_test.cpp (hunk 6 above)
Comment (dkalinowski) [reply src/test/mediapipeflow_test.cpp]: Fixed

**Review by atobiszei — APPROVED** [review]

File: src/test/mediapipeflow_test.cpp [hunk 7]
```diff
@@ -975,6 +1006,402 @@ TEST_P(MediapipeFlowAddTest, Infer) {
...
+            return true;
+        });
+    auto status = impl.ModelStreamInferImpl(nullptr, &stream);
+    ASSERT_EQ(status, StatusCode::OK) << status.string();
+}
+
+// Inference on unloaded mediapipe graph
+// Expect old stream to continue responding until closure
+// Expect new stream to be rejected
+TEST_P(MediapipeFlowAddTest, InferStreamOnUnloadedGraph) {
```
[the remark this answers is not in the document]
Comment (dkalinowski) [reply src/test/mediapipeflow_test.cpp]: Done

**Pull request merged**